File: templates/typescript-tests/src/workflows/generator.ts
import { Task, Workflow, OlapTable, Key } from "@514labs/moose-lib";
import { Foo, FooPipeline } from "../ingest/models";
import { faker } from "@faker-js/faker";

// Data model for OLAP Table
interface FooWorkflow {
  id: Key<string>;
  success: boolean;
  message: string;
}

// Create OLAP Table
const workflowTable = new OlapTable<FooWorkflow>("FooWorkflow");

export const ingest = new Task<null, void>("ingest", {
  run: async () => {
    // Use three fixed timestamps for E2E tests to add variability
    // while ensuring predictable results for consumption API tests
    const timestamps = [
      1739865600000, // Feb 18, 2025 00:00:00 UTC (day 18 - should NOT appear in day 19 queries)
      1739952000000, // Feb 19, 2025 00:00:00 UTC (day 19 - the target day for tests)
      1740038400000, // Feb 20, 2025 00:00:00 UTC (day 20 - should NOT appear in day 19 queries)
    ];

    for (let i = 0; i < 1000; i++) {
      // Cycle through the three timestamps to distribute data across days
      // This tests that aggregation and filtering work correctly
      const baseTimestamp = timestamps[i % 3];
      const fooHttp: Foo = {
        primaryKey: faker.string.uuid(),
        timestamp: baseTimestamp,
        optionalText:
          Math.random() < 0.5 ? "from_http\n" + faker.lorem.text() : undefined,
      };

      const fooSend: Foo = {
        primaryKey: faker.string.uuid(),
        timestamp: baseTimestamp,
        optionalText:
          Math.random() < 0.5 ? "from_send:\n" + faker.lorem.text() : undefined,
      };

      // HTTP ingest path
      try {
        const response = await fetch("http://localhost:4000/ingest/Foo", {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(fooHttp),
        });
        if (!response.ok) {
          workflowTable.insert([
            {
              id: "1",
              success: false,
              message: `HTTP ${response.status} ${response.statusText}`,
            },
          ]);
        }
      } catch (error) {
        workflowTable.insert([
          {
            id: "1",
            success: false,
            message: `HTTP error: ${(error as Error).message}`,
          },
        ]);
      }

      // Direct stream send path
      try {
        await FooPipeline.stream!.send(fooSend);
      } catch (error) {
        workflowTable.insert([
          {
            id: "1",
            success: false,
            message: `SEND error: ${(error as Error).message}`,
          },
        ]);
      }

      // Add a small delay to avoid overwhelming the server
      if (i % 100 === 0) {
        console.log(`Ingested ${i} records...`);
        workflowTable.insert([
          { id: "1", success: true, message: `Ingested ${i} records` },
        ]);
        await new Promise((resolve) => setTimeout(resolve, 100));
      }
    }
  },
  retries: 3,
  timeout: "30s",
});

export const workflow = new Workflow("generator", {
  startingTask: ingest,
  retries: 3,
  timeout: "30s",
  // schedule: "@every 5s",
});

File: templates/python-tests/src/workflows/generator.py
from moose_lib import Task, TaskConfig, Workflow, WorkflowConfig, OlapTable, InsertOptions, Key, TaskContext
from pydantic import BaseModel
from datetime import datetime
from faker import Faker
from src.ingest.models import Foo, Baz, fooModel
import requests

class FooWorkflow(BaseModel):
    id: Key[str]
    success: bool
    message: str

workflow_table = OlapTable[FooWorkflow]("foo_workflow")

def run_task(ctx: TaskContext[None]) -> None:
    fake = Faker()
    # Use three fixed timestamps for E2E tests to add variability
    # while ensuring predictable results for consumption API tests
    timestamps = [
        1739865600000,  # Feb 18, 2025 00:00:00 UTC (day 18 - should NOT appear in day 19 queries)
        1739952000000,  # Feb 19, 2025 00:00:00 UTC (day 19 - the target day for tests)
        1740038400000,  # Feb 20, 2025 00:00:00 UTC (day 20 - should NOT appear in day 19 queries)
    ]

    for i in range(1000):
        # Cycle through the three timestamps to distribute data across days
        # This tests that aggregation and filtering work correctly
        base_ts = timestamps[i % 3]

        # HTTP path payload
        foo_http = Foo(
            primary_key=fake.uuid4(),
            timestamp=base_ts,
            baz=fake.random_element(Baz),
            optional_text=("from_http\n" + fake.text()) if fake.boolean() else None,
        )

        # Direct send payload
        foo_send = Foo(
            primary_key=fake.uuid4(),
            timestamp=base_ts,
            baz=fake.random_element(Baz),
            optional_text=("from_send\n" + fake.text()) if fake.boolean() else None,
        )

        # HTTP ingest path
        try:
            req = requests.post(
                "http://localhost:4000/ingest/Foo",
                data=foo_http.model_dump_json().encode('utf-8'),
                headers={'Content-Type': 'application/json'}
            )
            if req.status_code == 200:
                workflow_table.insert([{"id": "1", "success": True, "message": f"HTTP inserted: {foo_http.primary_key}"}])
            else:
                workflow_table.insert([{"id": "1", "success": False, "message": f"HTTP failed: {req.status_code}"}])
        except Exception as e:
            workflow_table.insert([{"id": "1", "success": False, "message": f"HTTP error: {e}"}])

        # Direct stream send path
        try:
            fooModel.get_stream().send(foo_send)
            workflow_table.insert([{"id": "1", "success": True, "message": f"SEND inserted: {foo_send.primary_key}"}])
        except Exception as e:
            workflow_table.insert([{"id": "1", "success": False, "message": f"SEND error: {e}"}])

ingest_task = Task[None, None](
    name="task",
    config=TaskConfig(run=run_task)
)

ingest_workflow = Workflow(
    name="generator",
    config=WorkflowConfig(
        starting_task=ingest_task,
        retries=3,
        timeout="30s",
        # uncomment if you want to run it automatically on a schedule
        # schedule="@every 5s",
    )
)
